ksp_plugin/plugin.cpp

```diff
@@ -393,61 +433,116 @@ void Plugin::SetVesselStateOffset(
 }
 
-void Plugin::FreeVesselsAndCollectPileUps() {
+void Plugin::FreeVesselsAndPartsAndCollectPileUps() {
   CHECK(!initializing_);
-  FreeVessels();
-  for (auto const& pair : vessels_) {
-    Vessel& vessel = *pair.second;
-    Subset<Vessel>::Find(vessel).mutable_properties().Collect(&pile_ups_);
+
+  for (auto it = vessels_.cbegin(); it != vessels_.cend();) {
+    not_null<Vessel*> vessel = it->second.get();
+    if (kept_vessels_.erase(vessel)) {
+      ++it;
+    } else {
+      CHECK(!is_loaded(vessel));
+      LOG(INFO) << "Removing vessel with GUID " << it->first;
+      it = vessels_.erase(it);
+    }
+  }
+  // Free old parts, and bind vessels.
+  for (not_null<Vessel*> const vessel : loaded_vessels_) {
+    vessel->FreeParts();
+    Part& first_part = *vessel->parts().begin()->second;
+    for (auto const& pair : vessel->parts()) {
+      Part& part = *pair.second;
+      Subset<Part>::Unite(Subset<Part>::Find(first_part),
+                          Subset<Part>::Find(part));
+    }
+  }
+  // We only need to collect one part per vessel, since the other parts are part
+  // of the same subset.
+  for (not_null<Vessel*> const vessel : loaded_vessels_) {
+    Part& first_part = *vessel->parts().begin()->second;
+    Subset<Part>::Find(first_part).mutable_properties().Collect(&pile_ups_,
+                                                                current_time_);
   }
 }
 
-void Plugin::AddPileUpToBubble(std::list<PileUp>::iterator const pile_up) {
-  pile_ups_in_bubble_.push_back(pile_up);
+void Plugin::SetPartApparentDegreesOfFreedom(
+    PartId const part_id,
+    DegreesOfFreedom<World> const& degrees_of_freedom) {
+  RigidMotion<World, ApparentBubble> world_to_apparent_bubble{
+      RigidTransformation<World, ApparentBubble>{
+          World::origin,
+          ApparentBubble::origin,
+          OrthogonalMap<Barycentric, ApparentBubble>::Identity() *
+              WorldToBarycentric()},
+      AngularVelocity<World>{},
+      Velocity<World>{}};
+  not_null<Vessel*> vessel = FindOrDie(part_id_to_vessel_, part_id);
+  CHECK(is_loaded(vessel));
+  not_null<Part*> const part = vessel->part(part_id);
+  CHECK(part->is_piled_up());
+  part->containing_pile_up()->iterator()->SetPartApparentDegreesOfFreedom(
+      part, world_to_apparent_bubble(degrees_of_freedom));
 }
 
 void Plugin::AdvanceTime(Instant const& t, Angle const& planetarium_rotation) {
   VLOG(1) << __FUNCTION__ << '\n'
           << NAMED(t) << '\n' << NAMED(planetarium_rotation);
   CHECK(!initializing_);
   CHECK_GT(t, current_time_);
-  ephemeris_->Prolong(t);
 
-  if (pile_ups_in_bubble_.empty()) {
+  ephemeris_->Prolong(t);
+  for (PileUp& pile_up : pile_ups_) {
+    pile_up.DeformPileUpIfNeeded();
+    pile_up.AdvanceTime(*ephemeris_,
+                        t,
+                        DefaultHistoryParameters(),
+                        DefaultProlongationParameters());
+    // NOTE(egg): now that |NudgeParts| doesn't need the bubble barycentre
```

Comment:
Yeah!  Make this a TODO, I noticed the same thing (and in some cases NudgePart involves the exact same computations as appending to the tail).